Reflected aria attributes containing element references

hamishwillee opened the pull request:
FF136 adds support for a lot of aria reflected elements that reference other elements. [Out-of-date explainer](https://wicg.github.io/aom/aria-reflection-explainer.html#reflecting-element-references).

NOTE FOR REVIEWER:
- These are attributes that contain id references to other elements. 
- They don't just say "reflects the attribute" because they don't. 
  - The attribute is a reference id or string of ids, but these ids might not map to actual elements, or might map to out of scope elements.
  - The property is an array of elements, which will take the valid values from the attribute if it is set.
  - If you set the property you clear the attribute. Last setter wins.

Because there is a lot to explain, that would otherwise have to be duplicated: 
- I have created [Document_Object_Model/Reflected_attributes ](https://pr38678.review.mdn.allizom.net/en-US/docs/Web/API/Document_Object_Model/Reflected_attributes) to cover reflected attributes, including this special case. That has examples that show the rules for scoping etc
- Each individual doc summarises those rules and cross links to them.
- Element property docs have a simple example that shows how the attribute might be used, and shows how the property is read (not as good as showing how the property might be used instead of the attribute - but simple and enough to get the idea)
- `ElementInternals` docs are _mostly_ a duplicate of the `Element` docs but with links to the examples in corresponding Element doc and the reflected attributes. Didn't seem worth duplicating the examples for what they show.

One thing I am still considering doing is another example in  [Document_Object_Model/Reflected_attributes ](https://pr38678.review.mdn.allizom.net/en-US/docs/Web/API/Document_Object_Model/Reflected_attributes)  that shows how a shadow root can access an element from the DOM.

Docs

- [x] Element.ariaActiveDescendantElement
- [x] Element.ariaControlsElements
- [x] Element.ariaDescribedByElements
- [x] Element.ariaDetailsElements
- [x] Element.ariaErrorMessageElements - in prog
- [x] Element.ariaFlowToElements
- [x] Element.ariaLabelledByElements - in progress
- [x] Element.ariaOwnsElements
- [x] ElementInternals.ariaActiveDescendantElement
- [x] ElementInternals.ariaControlsElements
- [x] ElementInternals.ariaDescribedByElements
- [x] ElementInternals.ariaDetailsElements
- [x] ElementInternals.ariaErrorMessageElements
- [x] ElementInternals.ariaFlowToElements
- [x] ElementInternals.ariaLabelledByElements  - in progress
- [x] ElementInternals.ariaOwnsElements
- [x] Element - top level ref links
- [x] ElementInternals - top level ref links
- [x] Reflecting attribute guide

Related docs can be tracked in https://github.com/mdn/content/issues/38402

hamishwillee committed:
ElementInternals/Element - reflected aria attributes containing element references

 variants#

hamishwillee commented on files/en-us/web/api/element/arialabelledbyelements/index.md:
```diff
@@ -0,0 +1,178 @@
+if ("ariaLabelledByElements" in Element.prototype) {
+  inputElement.ariaLabelledByElements = [label1Element, label2Element];
+
+  log(`Labeled elements: ${inputElement.getAttribute("aria-labelledby")}`);
```
According to explainer we should be able to set the `ariaLabelledByElements` and get back the references using the `getAttribute()` - provided they have ids. This doesn't work yet

hamishwillee replied:
This is not how it works!

hamishwillee commented on files/en-us/web/api/elementinternals/arialabelledbyelements/index.md:
```diff
@@ -0,0 +1,356 @@
+
+{{APIRef("DOM")}}
+
+The **`ariaLabelledByElements`** property of the {{domxref("ElementInternals")}} interface reflects the value of the [`aria-labelledby`](/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-labelledby) attribute, which identifies the element (or elements) that provide an accessible name for the element it is applied to.
```
At this point this is a duplicate of the equivalent element property but it assigns the aria relationship within the shadow root. Not yet sure what else is useful to show.

hamishwillee commented on files/en-us/web/api/elementinternals/arialabelledbyelements/index.md:
```diff
@@ -0,0 +1,356 @@
+
+{{EmbedLiveSample("Get the accessible name from other elements","100%","150px")}}
+
+<!-- Below is test code - shows that ariaLabelledByElements isn't picking up the label that is assigned outside the shadow
```
Note, this is (hidden) same example, but which has one of the elements outside the shadow root - and this is not listed by ariaLabelledByElements. So it may be that there is no selection outside of the boundary.

hamishwillee commented on files/en-us/web/api/element/ariaactivedescendantelement/index.md:
```diff
@@ -0,0 +1,274 @@
+if ("ariaActiveDescendantElement" in Element.prototype) {
+  // Log id for active descendant relationship through the content attribute.
+  log(
+    `aria-activedescendant (id): ${streetType.getAttribute("aria-activedescendant")}`,
```
Note, all these examples show the id from the attribute as empty. I'm leaving it here for now because the [explainer](https://wicg.github.io/aom/aria-reflection-explainer.html#reflecting-element-references) says this should be populated - "sprouted" - when you set the attribute.

hamishwillee commented on files/en-us/web/api/element/ariaactivedescendantelement/index.md:
```diff
@@ -0,0 +1,274 @@
+  // Our active descendant relationship remained intact!
+  log(
+    `txt: ${streetType.ariaActiveDescendantElement?.textContent.trim()}, el: ${streetType.ariaActiveDescendantElement}, id: ${streetType.getAttribute("aria-activedescendant")}`,
+  );
```
This all works, but I am a bit confused by whether you should be able to set an active descendant for a DOM element to be inside shadow DOM or not. Testing seems to say no?

hamishwillee replied:
Note, you can't. You can set in same scope or ancestor scope.

hamishwillee commented on files/en-us/web/api/element/ariaactivedescendantelement/index.md:
```diff
@@ -0,0 +1,201 @@
+
+{{EmbedLiveSample("Get the active descendent","100%","190px")}}
+
+<!-- Come back to this. Show shadow root can't be selected, and also this, that you can move in and out of scope -->
+<!--
+
```
Bit below here to be redone.

hamishwillee commented on files/en-us/web/api/element/ariaactivedescendantelement/index.md:
```diff
@@ -0,0 +1,203 @@
+An element that is the active descendant, or `null` if there is no active descendant.
+
+Note that when the property is set, the corresponding [`aria-activedescendant`](/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-activedescendant) attribute is set to the empty string (`""`), and this will be returned by {{domxref("Element.getAttribute()")}}.
+The attribute can subsequently be set using {{domxref("Element.setAttribute()")}}, in which case the referenced element will replace the property value if it is in scope.
```
```suggestion
The attribute can subsequently be set using {{domxref("Element.setAttribute()")}}, in which case the referenced element will replace the property value if it is in scope, or be set to `null` if it is not in scope/invalid.
```

hamishwillee committed:
Add links to associated interfaces from the attributes

bsmth committed:
chore(aria): please the formatter

hamishwillee commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
```diff
@@ -0,0 +1,420 @@
+  If you read the attribute it will return `""`.
+- Setting the attribute (with {{domxref("Element.setAttribute()")}}) causes the property to reflect the valid and in-scope references.
+- Setting the attribute with a value reference that is subsequently moved out of scope will result in removal of the corresponding element from the property array.
+  Note however that the attribute still contains the reference, and if the element is moved back in-scope the property will again include the element (i.e. the relationship is restored).
```
I hate you review dog. While technically correct, the addition of a comma hare adds zero value.

hamishwillee committed:
Update files/en-us/web/api/document_object_model/reflected_attributes/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

bsmth commented on files/en-us/glossary/attribute/index.md:
```diff
@@ -15,33 +15,16 @@ A number of HTML attributes are {{Glossary("Boolean/HTML", "boolean attributes")
-and can be modified by setting the property to a different value.
 
-For example, the `placeholder` below is reflected into {{domxref("HTMLInputElement.placeholder")}}.
+What this means is that the value of the attribute can be read or written directly in JavaScript through a property on the corresponding interface, and vice versa.
```
```suggestion
This means that the value of the attribute can be read or written directly in JavaScript through a property on the corresponding interface, and vice versa.
```

bsmth commented on files/en-us/web/accessibility/aria/reference/attributes/aria-activedescendant/index.md:
```diff
@@ -37,6 +37,11 @@ The value of `aria-activedescendant` refers to an owned element of the controlle
+## Associated interfaces
+
+- {{domxref("Element.ariaActiveDescendantElement")}} and {{domxref("ElementInternals.ariaActiveDescendantElement")}}
+  - : The `ariaActiveDescendantElement` property is part of each element's interface. Its value is an {{domxref("Element")}}s corresponding to the ID value specified in the `aria-activedescendant` attribute (for cases where the ID values references a valid and in-scope element).
```
typo?

```suggestion
  - : The `ariaActiveDescendantElement` property is part of each element's interface. Its value is an {{domxref("Element")}} corresponding to the ID value specified in the `aria-activedescendant` attribute (for cases where the ID values references a valid and in-scope element).
```

wbamberg replied:
I think it would be clearer to have separate `dt/dd` pairs for `Element` and `ElementInternals`. It was for me hard even to recognize that this was a dl: because the dt wraps, it looks like just a couple of paragraphs.

It would be fine for the `ElementInternals` one to say basically, the same but for the `ElementInternals` interface.

> (for cases where the ID values references a valid and in-scope element).

...feels like it's packing a lot of meaning into too few words. I mean: if I don't already know what this means, this isn't enough to understand it.

bsmth commented on files/en-us/web/accessibility/aria/reference/attributes/aria-errormessage/index.md:
```diff
@@ -6,15 +6,15 @@ spec-urls: https://w3c.github.io/aria/#aria-errormessage
 ## Description
 
-When there is a user-created error, you want to let them know it exists and tell them how to fix it. There are two attributes you need to use: set [`aria-invalid="true"`](/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-invalid) to define the object as being in an error state, then add the `aria-errormessage` attribute with the value being the `id` of the element containing the error  ...
+When there is a user-created error, you want to let them know it exists and tell them how to fix it. There are two attributes you need to use: set [`aria-invalid="true"`](/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-invalid) to define the object as being in an error state, then add the `aria-errormessage` attribute with the value being the `id` of the element (or elements) contain ...
```
```suggestion
When there is a user-created error, you want to let the user know it exists and tell them how to fix it. There are two attributes you need to use: set [`aria-invalid="true"`](/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-invalid) to define the object as being in an error state, then add the `aria-errormessage` attribute with the value being the `id` of the element (or elements) containing the error message text for that object.
```

bsmth commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
```diff
@@ -0,0 +1,420 @@
+
+{{DefaultAPISidebar("DOM")}}
+
+An {{Glossary("attribute")}} extends an {{Glossary("HTML")}}, {{Glossary("XML")}}, {{Glossary("SVG")}} or other {{Glossary("element")}}, changing its behavior or providing metadata.
```
We can do lowercase while we're at it if you want:

```suggestion
An {{glossary("attribute")}} extends an {{glossary("HTML")}}, {{glossary("XML")}}, {{glossary("SVG")}} or other {{glossary("element")}}, changing its behavior or providing metadata.
```

bsmth commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
```diff
@@ -0,0 +1,420 @@
+An {{Glossary("attribute")}} extends an {{Glossary("HTML")}}, {{Glossary("XML")}}, {{Glossary("SVG")}} or other {{Glossary("element")}}, changing its behavior or providing metadata.
+
+Many attributes are _reflected_ in the corresponding [DOM](/en-US/docs/Web/API/Document_Object_Model) interface.
+What this means is that the value of the attribute can be read or written directly in JavaScript through a property on the corresponding interface, and vice versa.
```
Same minor suggestion:

```suggestion
This means that the value of the attribute can be read or written directly in JavaScript through a property on the corresponding interface, and vice versa.
```

bsmth commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
````diff
@@ -0,0 +1,420 @@
+```
+
+Note that the name of the reflected attribute and the property are the same: `placeholder`.
+This is not always the case: properties are usually named following the {{glossary("Camel case","camelCase")}} convention.
````
Don't we talk about this conversion algorithm somewhere? (`aria-checked` to `ariaChecked`)

hamishwillee replied:
Not that I know of. Don't see as necessary to link as it is obvious.

bsmth replied:
Righto

bsmth commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
```diff
@@ -0,0 +1,420 @@
+- The property may be assigned an element that does not have an `id`, and which therefore can't be referenced in the attribute (being able to assign elements without having to create an id is actually a benefit of using properties!)
+
+For these reasons, unlike for other reflected elements, there may not be a 1:1 correspondence between the reflected attributes and their associated property.
+The way that these issues are resolved is that:
```
```suggestion
These issues are resolved as follows:
```

bsmth replied:
A little shorter if you like it, or "These issues are resolved like so:"

bsmth commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
```diff
@@ -0,0 +1,420 @@
+  Note however that the attribute still contains the reference, and if the element is moved back in-scope the property will again include the element (i.e., the relationship is restored).
+
+Note that the allowed scope of a reference is an element in the same scope or an ancestor scope of the element, but not a descendant scope.
+What this means that an element in a shadow root can reference an element from within its own shadow DOM or the parent DOM, but a DOM element can't set an element defined in a (descendent) shadow root.
```
```suggestion
This means that an element in a shadow root can reference an element from within its own shadow DOM or the parent DOM, but a DOM element can't set an element defined in a (descendent) shadow root.
```

bsmth commented on files/en-us/web/api/element/ariacontrolselements/index.md:
```diff
@@ -0,0 +1,150 @@
+
+- [`aria-controls`](/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-controls) attribute
+- {{domxref("ElementInternals.ariaControlsElements")}}
+- [Reflected element references](2/en-US/docs/Web/API/Document_Object_Model/Reflected_attributes#reflected_element_references) in the _Attribute reflection_ guide
```
Maybe check for other places this might be:

```suggestion
- [Reflected element references](/en-US/docs/Web/API/Document_Object_Model/Reflected_attributes#reflected_element_references) in the _Attribute reflection_ guide
```

bsmth replied:
It's in a few other docs, so I won't point it out everywhere.

hamishwillee replied:
Thanks. I'll do a check after merging this.

bsmth commented on files/en-us/web/api/element/ariaerrormessageelements/index.md:
````diff
@@ -0,0 +1,147 @@
+```js
+const email = document.querySelector("#email");
+
+email.addEventListener("input", (event) => {
````
Just a general note that we don't have event listeners on other / all examples that have an `<input>` (like `ariaLabelledByElements`), so I would mention on those that the reader shouldn't expect that it does something when they type some text in there.

hamishwillee replied:
Good point. I have tried to address this in https://github.com/mdn/content/pull/38678/commits/93b5053f4ea537305c0be9f41d5eb936a0948f05

Note, I might modify the examples tomorrow to make this point moot. Nearly there on this.

wbamberg commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
```diff
@@ -0,0 +1,420 @@
+function logAccessibleInfo(element) {
+  const refids = element.getAttribute("aria-labelledby");
+
+  let text = "";
+  accessElements = element.ariaLabelledByElements;
+  accessElements.forEach((el) => {
+    text += el.textContent.trim() + " ";
+  });
+  text = text.trim();
```
```suggestion
  const text = element.ariaLabelledByElements.map((e) => e.textContent).join(" ");
```

...here and elsewhere.

hamishwillee replied:
I believe I have fixed this everywhere in https://github.com/mdn/content/pull/38678/commits/29b8f88ea3280b219cba659a347bc323874520e9

wbamberg commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
```diff
@@ -0,0 +1,420 @@
+
+### Reflected element reference DOM scope
+
+This example demonstrates that an element can't reference a descendent scope, but that a referenced that is moved out of scope will be reflected again when it is moved back into scope.
```
```suggestion
This example demonstrates that an element can't reference an element in a descendent scope, but that a referenced element that is moved out of scope will be reflected again when it is moved back into scope.
```

...but as above I'd like to know more about these definitions.

wbamberg reviewed (commented):
Some comments on the guide page.

wbamberg commented on files/en-us/web/api/element/ariacontrolselements/index.md:
```diff
@@ -0,0 +1,150 @@
+
+{{APIRef("DOM")}}
+
+The **`ariaControlsElements`** property of the {{domxref("Element")}} interface is an array containing the element (or elements) that are controlled by the element it is applied to.
```
```suggestion
The **`ariaControlsElements`** property of the {{domxref("Element")}} interface is an array containing the elements that are controlled by the element it is applied to.
```

picky but it's an obvious simplification.

wbamberg replied:
This change doesn't seem to be made?

wbamberg replied:
This change doesn't seem to be made?

wbamberg commented on files/en-us/web/api/element/ariacontrolselements/index.md:
```diff
@@ -0,0 +1,150 @@
+
+## Value
+
+An array of elements that are controlled by this element.
```
```suggestion
An array of subclasses of {{domxref("HTMLElement")}}, representing the elements that are controlled by this element.
```

wbamberg commented on files/en-us/web/api/element/ariacontrolselements/index.md:
```diff
@@ -0,0 +1,150 @@
+## Description
+
+The property is a flexible alternative to using the [`aria-controls`](/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-controls) attribute to set the controlled elements.
+Unlike `aria-controls`, the elements assigned to this property do not have to have an `id`: they can be selected.
```
Same comment as in the other page.

hamishwillee replied:
```suggestion
Unlike `aria-controls`, the elements assigned to this property do not have to have an [`id`](/en-US/docs/Web/HTML/Global_attributes/id) attribute.
```

wbamberg commented on files/en-us/web/api/element/ariacontrolselements/index.md:
```diff
@@ -0,0 +1,150 @@
+
+The property is a flexible alternative to using the [`aria-controls`](/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-controls) attribute to set the controlled elements.
+Unlike `aria-controls`, the elements assigned to this property do not have to have an `id`: they can be selected.
+This can be convenient as it avoids having to unnecessarily create ids for elements in order to set them as controlled.
```
This sentence isn't present in the other page although it applies just as much there. I thought it was an odd expression: the id attributes are not unnecessary, they are necessary exactly for this purpose.

tbh I don't think this sentence is needed, the implication of the previous one is surely obvious.

wbamberg commented on files/en-us/web/api/element/ariacontrolselements/index.md:
```diff
@@ -0,0 +1,150 @@
+This can be convenient as it avoids having to unnecessarily create ids for elements in order to set them as controlled.
+
+The property reflects the [`aria-controls`](/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-controls) attribute when it is defined, but only for listed reference `id` values that match valid in-scope elements.
+If the property is set, then the corresponding attribute is cleared.
```
Question, I forgot to ask about the guide page: this is an array, what happens to the attribute if I append or delete an element, is it still cleared?

hamishwillee replied:
OK, so the property appears to return a read-only fixed array. If you try to add an element to the array you get a TypeError. If you update the attribute the previously returned array is not modified, so you'd have to read the property again to get the current value. 
In other words, not possible to update the array, you can only reassign a new value to the property. I will modify these everywhere to note that a read-only array is returned.

wbamberg commented on files/en-us/web/api/element/ariacontrolselements/index.md:
```diff
@@ -0,0 +1,150 @@
+  controlledElements.forEach((controlled) => {
+    textOfControllingElements += `"${controlled.textContent.trim()}" `;
+  });
+  log(`Controlled element text: ${textOfControllingElements.trim()}`);
```
As in the guide page, it would be really nice to avoid the horizontal scroll here.

hamishwillee replied:
I think I have fixed this in https://github.com/mdn/content/pull/38678/commits/29b8f88ea3280b219cba659a347bc323874520e9

wbamberg commented on files/en-us/web/accessibility/aria/reference/attributes/aria-controls/index.md:
```diff
@@ -84,8 +84,8 @@ In this tabs example, each tab controls one tabpanel:
-- {{domxref("Element.ariaControlsElements")}}
-  - : The `ariaControlsElements` property is part of each element's interface. Its value is a list of {{domxref("Element")}}s corresponding to the ID values specified in the `aria-controls` attribute.
+- {{domxref("Element.ariaControlsElements")}} and {{domxref("ElementInternals.ariaControlsElements")}}
+  - : The `ariaControlsElements` property is part of each element's interface. Its value is an array of {{domxref("Element")}}s corresponding to the ID values specified in the `aria-controls` attribute (for cases where those ID values reference valid and in-scope elements).
```
Is it only me that finds usages of identifier names mixed into prose grating? Like "An array of `Element`s", rather than "An array of `Element` instances"?

hamishwillee replied:
No, I find it very grating, at least visually. 

> What does "valid" mean here? "valid and in-scope elements" implies that we have 2 conditions, but what would it look like for an element to be in scope but not valid?

Good point. Valid means "exists" - i.e. I can reference ID "fred" but if no element with `id="fred"`. I think I can just use valid to mean "must exist and be in scope". These will then become:

```suggestion
  - : The `ariaControlsElements` property is part of each element's interface. Its value is an array of {{domxref("Element")}} instances corresponding to the valid ID values specified in the `aria-controls` attribute.
```

wbamberg reviewed (commented):
As you asked, I just looked at a few sample pages.

hamishwillee commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
```diff
@@ -0,0 +1,420 @@
+- Setting the attribute with a value reference that is subsequently moved out of scope will result in removal of the corresponding element from the property array.
+  Note however that the attribute still contains the reference, and if the element is moved back in-scope the property will again include the element (i.e., the relationship is restored).
+
+Note that the allowed scope of a reference is an element in the same scope or an ancestor scope of the element, but not a descendant scope.
```
The terms came out of a question I had, and are not AFAIK used in the spec. But I am not sure so have asked in https://bugzilla.mozilla.org/show_bug.cgi?id=1919102#c18. Either way, I think I am happy to define them here.

Yes, the definition is in the next sentence. I take the "scope" of an element to mean the current DOM/ShadowDOM tree in which it is defined. An ancestor scope is a DOM tree that nests the element's DOM, and a descendent scope is a shadowDOM nested by the element's DOM.

Something is in-scope with the element w.r.t. element references if it is in the parent scope or an ancestor scope, but not a descendent scope.

What that means is that an element in a document can't reference the ids of elements in a nested shadowDOM, because they are descendants. But an element in the nested shadow dom of that same document could reference the id of the parent. 

This fell out of https://bugzilla.mozilla.org/show_bug.cgi?id=1919102#c14

I might leave addressing this until I get back from holiday and you've had a chance to think about this response.

wbamberg replied:
I think what I'm trying to get at is something like this. Sorry if this is obvious.

Say, there are APIs like `getElementById()` or `querySelector("#thing")`. These have rules about when they will find an element with the given ID - it has to be in the document, obviously, etc. Question is, is the "in scope" here using the same rules? That is, is this just a specific application of a more general rule. If it is, then it should be documented somewhere else, for instance the `id` page, and just referenced here.

> I might leave addressing this until I get back from holiday and you've had a chance to think about this response.

Sure, sorry to give you all this feedback today! Please choose what you want to do with it :).

hamishwillee commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
```diff
@@ -0,0 +1,420 @@
+
+## Examples
+
+### Getting and setting reflected attributes
```
Yes. Grrr.

hamishwillee commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
```diff
@@ -0,0 +1,420 @@
+
+{{EmbedLiveSample("Getting and setting reflected attributes","100%","350px")}}
+
+### Setting and getting reflected element references
```
Agree. Nothing seems like an elegant way of doing this.  

It might be better if I triggered the parts of the logs on buttons - so something like "Add invalid reference" button that you press, and the log shows the reference that was added and the lack of similar reference appearing and an explanation.

hamishwillee commented on files/en-us/web/api/element/ariaactivedescendantelement/index.md:
```diff
@@ -0,0 +1,105 @@
+## Description
+
+The property is a flexible alternative to using the [`aria-activedescendant`](/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-activedescendant) attribute.
+Unlike `aria-activedescendant`, the element assigned to this property can be selected: it does not have to have an `id`.
```
Thanks. This is indeed the point.

hamishwillee committed:
Apply suggestions from code review - easy

Co-authored-by: Brian Smith <brian@smith.berlin>
Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee committed:
Fix the broken links starting with 2/en-Us

hamishwillee committed:
Apply suggestions from code review

Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee committed:
Value - reference the types of the object

hamishwillee committed:
Link id global and remove obvious inference

hamishwillee commented on files/en-us/web/api/element/ariaactivedescendantelement/index.md:
```diff
@@ -0,0 +1,105 @@
+The property is a flexible alternative to using the [`aria-activedescendant`](/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-activedescendant) attribute.
+Unlike `aria-activedescendant`, the element assigned to this property can be selected: it does not have to have an `id`.
+
+The property reflects the element's [`aria-activedescendant`](/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-errormessage) attribute when it is defined, but only for reference `id` values that match valid in-scope elements.
```
It is logical - if `getAttribute()` returns a non-empty value, then the property will have an array containing the elements that have matching `id` values in the same DOM or an ancestor (parent) DOM.

This is a bit tricky, because when I think "attribute" my first though is "the text applied to the tag" - but of course it is actually another dynamic property.

I will roll out this pattern change everywhere:

```suggestion
The property reflects the element's [`aria-activedescendant`](/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-errormessage) attribute if it is has a value other than the empty string, but only for listed reference `id` values that match elements in this element's DOM tree, or a parent DOM tree.
```

hamishwillee replied:
Marking as resolved as we talked this through offline.

hamishwillee committed:
Create an in-scope and link

hamishwillee commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
```diff
@@ -0,0 +1,435 @@
+
+The attribute's element references are reflected in the property as an array of the corresponding {{domxref("HTMLElement")}}-derived object instances, with some caveats.
+
+### Element reference scope
+
+Element references can only match to elements that are "in scope" with the referencing element.
+
+An HTML document is represented to JavaScript as a hierarchical tree of objects referred to as the [Document Object Model (DOM)](/en-US/docs/Web/API/Document_Object_Model).
+Elements within the model may contain and encapsulate "child" DOMs, referred to as [Shadow DOMs](/en-US/docs/Web/API/Web_components#shadow_dom_2) within a [`ShadowRoot`](/en-US/docs/Web/API/ShadowRoot), which can in turn nest their own Shadow DOMs.
+
+The scope of a referencing element is the DOM in which it is defined, and any parent DOMs in which that DOM is nested. Shadow DOMS that are nested children of the DOM in which the referencing element is defined are out of scope.
+
+What this means is that an element in the main DOM can only reference elements in the main DOM (since elements in a Shadow DOM are necessarily children of its DOM, and hence out of scope).
+However an element in a shadow root can nest elements in that same shadow root and also in the main DOM (and any DOM's that are parents), since they are in scope.
+
+<!-- diagrams? -->
+
+Note that a referenced element may initially be "in scope" and then moved out of scope into a nested shadow root.
+In this case the referenced element may still be listed in the attribute
```
This is my rewritten section for "in scope" which I hope to link from everywhere. I might add diagrams. 

These are other comments from @wbamberg that disappeared when I added (need to look at them still)

https://github.com/mdn/content/pull/38678/files#r2035809649
https://github.com/mdn/content/pull/38678/files#r2035809649

hamishwillee commented on files/en-us/web/accessibility/aria/reference/attributes/aria-flowto/index.md:
```diff
@@ -22,9 +22,19 @@ When `aria-flowto` has a single [id](/en-US/docs/Web/HTML/Global_attributes/id)
+
+## Associated interfaces
+
+- {{domxref("Element.ariaFlowToElements")}}
+  - : The `ariaFlowToElements` property is part of each element's interface.
+    Its value is an array of instances of subclasses of {{domxref("Element")}} that reflect the `id` references in the `aria-flowto` attribute ([with some caveats](/en-US/docs/Web/API/Document_Object_Model/Reflected_attributes#reflected_element_references)).
+- {{domxref("ElementInternals.ariaFlowToElements")}}
+  - : The `ariaFlowToElements` property is part of each custom element's interface.
+    Its value is an array of instances of subclasses of {{domxref("Element")}} that reflect the `id` references in the `aria-flowto` attribute ([with some caveats](/en-US/docs/Web/API/Document_Object_Model/Reflected_attributes#reflected_element_references)).
```
@wbamberg This is my latest attempt on the interfaces. As you suggested I split these into two DLs. However I decided not to explain the details of id reflection and scope. Instead I just added a link to that whole discussion on reflected element references.

My thinking is this. At this point you need to know that the corresponding interface exists, that it reflects the IDs, and that this might not be straightforward. Linking to the with some caveats does that. 

IF you think better, I could rename the link to "See [Reflected element reference]" - which might make sense

wbamberg replied:
I think this is a good idea. It's totally better to link to a full explanation rather than try to pack it into too small a space.

hamishwillee committed:
fix up the aria attr to separate interfaces

hamishwillee committed:
Note the returned value is static/read only

hamishwillee committed:
Fix example horiz scroll and simplify getting inner text

hamishwillee commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
```diff
@@ -0,0 +1,420 @@
+  This should show that `label_3` has no corresponding element reflected in the property, as it is out of scope.
+- As you toggle the button the `label_1` element should be added and removed from the property as it is moved in and out of scope.
+
+{{EmbedLiveSample("Reflected element reference DOM scope","100%","350px")}}
```
I think I have fixed these cases in https://github.com/mdn/content/pull/38678/commits/29b8f88ea3280b219cba659a347bc323874520e9

hamishwillee committed:
Add note about text entered into inputs being ignored

hamishwillee committed:
Make guide more guidy

hamishwillee commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
```diff
@@ -0,0 +1,420 @@
+For these reasons, unlike for other reflected elements, there may not be a 1:1 correspondence between the reflected attributes and their associated property.
+The way that these issues are resolved is that:
+
+- The attribute is only reflected when it is defined, and only for listed reference `id` values that match valid in-scope elements.
```
Thank you. I have attempted to follow this path

hamishwillee committed:
revert accidental change

hamishwillee commented on files/en-us/web/api/htmlinputelement/placeholder/index.md:
````diff
@@ -16,7 +16,7 @@ A string.
 ## Examples
 
-```js
+```
````
````suggestion
```js
````

hamishwillee committed:
typo

hamishwillee commented:
@bsmth I've addressed your issues, but made further updates following feedback from both you and WillB.

@wbamberg I've addressed your feedback too. Can you please sanity check updates, in particular to the [/en-US/docs/Web/API/Document_Object_Model/Reflected_attributes](https://pr38678.review.mdn.allizom.net/en-US/docs/Web/API/Document_Object_Model/Reflected_attributes)

bsmth commented on files/en-us/glossary/attribute/index.md:
```diff
@@ -15,33 +15,16 @@ A number of HTML attributes are {{Glossary("Boolean/HTML", "boolean attributes")
 
-For example, the `placeholder` below is reflected into {{domxref("HTMLInputElement.placeholder")}}.
+This means that the value of the attribute can be read or written directly in JavaScript through a property on the corresponding interface, and vice versa.
+The reflected properties offer a more natural programming approach than getting and setting attribute using the {{domxref("Element.getAttribute()","getAttribute()")}} and {{domxref("Element.setAttribute()","setAttribute()")}} methods of the {{domxref("Element")}} interface.
```
Small one:

```suggestion
The reflected properties offer a more natural programming approach than getting and setting attributes using the {{domxref("Element.getAttribute()","getAttribute()")}} and {{domxref("Element.setAttribute()","setAttribute()")}} methods of the {{domxref("Element")}} interface.
```

bsmth reviewed (approved):
One small comment, but otherwise it's looking good to me, thanks! Passing it over to Will 🙏🏻

wbamberg commented on files/en-us/web/api/element/ariacontrolselements/index.md:
```diff
@@ -0,0 +1,148 @@
+
+An array of subclasses of {{domxref("HTMLElement")}}, representing the elements that are controlled by this element.
+
+When read, the returned array is a static and read-only.
```
```suggestion
When read, the returned array is static and read-only.
```

? or even just "read-only"?

hamishwillee replied:
My understanding is that "it is possible" that in general you might be handed an array that you can't modify, but that might still be modified by the supplier - i.e. it might be read only but still updated if the attribute changes. 

However this is static - no one can modify it after creation.

It may be that I am wrong and that read only always means static in JavaScript?

wbamberg commented on files/en-us/web/accessibility/aria/reference/attributes/aria-activedescendant/index.md:
```diff
@@ -37,6 +37,15 @@ The value of `aria-activedescendant` refers to an owned element of the controlle
+
+- {{domxref("Element.ariaActiveDescendantElement")}}
+  - : The `ariaActiveDescendantElement` property is part of each element's interface.
+    Its value is an array of instances of subclasses of {{domxref("Element")}} that reflect the `id` references in the `aria-activedescendant` attribute ([with some caveats](/en-US/docs/Web/API/Document_Object_Model/Reflected_attributes#reflected_element_references)).
```
This one I think is not an array.

hamishwillee replied:
Doh. Thanks. 

```suggestion
    Its value is an instance of a subclass of {{domxref("Element")}} that reflects the `id` reference in the `aria-activedescendant` attribute ([with some caveats](/en-US/docs/Web/API/Document_Object_Model/Reflected_attributes#reflected_element_references)).
```

wbamberg commented on files/en-us/web/accessibility/aria/reference/attributes/aria-activedescendant/index.md:
```diff
@@ -37,6 +37,15 @@ The value of `aria-activedescendant` refers to an owned element of the controlle
+    Its value is an array of instances of subclasses of {{domxref("Element")}} that reflect the `id` references in the `aria-activedescendant` attribute ([with some caveats](/en-US/docs/Web/API/Document_Object_Model/Reflected_attributes#reflected_element_references)).
+- {{domxref("ElementInternals.ariaActiveDescendantElement")}}
+  - : The `ariaActiveDescendantElement` property is part of each custom element's interface.
+    Its value is an array of instances of subclasses of {{domxref("Element")}} that reflect the `id` references in the `aria-activedescendant` attribute ([with some caveats](/en-US/docs/Web/API/Document_Object_Model/Reflected_attributes#reflected_element_references)).
```
Same.

hamishwillee replied:
```suggestion
    Its value is an instance of a subclass of {{domxref("Element")}} that reflects the `id` reference in the `aria-activedescendant` attribute ([with some caveats](/en-US/docs/Web/API/Document_Object_Model/Reflected_attributes#reflected_element_references)).
```

wbamberg commented on files/en-us/web/api/document_object_model/reflected_attributes/index.md:
```diff
@@ -0,0 +1,231 @@
+## Reflected element references
+
+> [!NOTE]
+> This section applies to [reflected aria attributes that contain element references](/en-US/docs/Web/API/Element#instance_properties_reflected_from_aria_element_references).
```
```suggestion
> This section applies to [reflected ARIA attributes that contain element references](/en-US/docs/Web/API/Element#instance_properties_reflected_from_aria_element_references).
```

wbamberg reviewed (commented):
Just a few little things.

hamishwillee commented on files/en-us/web/api/element/arialabelledbyelements/index.md:
```diff
@@ -0,0 +1,116 @@
+
+### Get the accessible name
+
+This example shows how `ariaLabelledByElements` can be used to get an aria label defined using `aria-labelledby`.
```
```suggestion
This example shows how `ariaLabelledByElements` can be used to get an ARIA label defined using `aria-labelledby`.
```

hamishwillee commented on files/en-us/web/api/element/index.md:
```diff
@@ -171,12 +171,40 @@ _The `Element` interface also includes the following properties._
 - {{domxref("Element.role")}}
   - : A string reflecting the explicitly set [`role`](/en-US/docs/Web/Accessibility/ARIA/Reference/Roles) attribute, which provides the semantic role of the element.
 
+#### Instance properties reflected from aria element references
```
```suggestion
#### Instance properties reflected from ARIA element references
```

hamishwillee commented on files/en-us/web/api/elementinternals/index.md:
```diff
@@ -125,6 +125,36 @@ The `ElementInternals` interface also includes the following properties.
 - {{domxref("ElementInternals.ariaValueText")}}
   - : A string reflecting the [`aria-valuetext`](/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-valuetext) attribute, which defines the human-readable text alternative of aria-valuenow for a range widget.
 
+#### Instance properties reflected from aria element references
```
```suggestion
#### Instance properties reflected from ARIA element references
```

hamishwillee committed:
Apply suggestions from code review

Co-authored-by: Brian Smith <brian@smith.berlin>
Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee commented:
Thank you very much @bsmth and @wbamberg for your reviews. Appreciate that this is not fun.
